src/cpp/Cdr.cpp

```diff
@@ -2754,10 +2879,919 @@ Cdr& Cdr::deserializeWStringSequence(
+{
+    static_cast<void>(current_state);
+    assert(current_state == Cdr::state(*this));
+    assert(EncodingAlgorithmFlag::PLAIN_CDR2 == current_encoding_ ||
+            EncodingAlgorithmFlag::DELIMIT_CDR2 == current_encoding_ ||
+            EncodingAlgorithmFlag::PL_CDR2 == current_encoding_);
+    assert(EncodingAlgorithmFlag::PLAIN_CDR2 == type_encoding ||
+            EncodingAlgorithmFlag::DELIMIT_CDR2 == type_encoding ||
+            EncodingAlgorithmFlag::PL_CDR2 == type_encoding);
+    assert(offset_ == m_cdrBuffer.begin() ? current_encoding_ == type_encoding : true);
+    if (EncodingAlgorithmFlag::PLAIN_CDR2 != type_encoding)
+    {
+        uint32_t dheader {0};
+        serialize(dheader);
+    }
+    serialized_member_size_ = NO_SERIALIZED_MEMBER_SIZE; // Avoid error when serializen arrays, sequences, etc..
+    current_state.previous_encoding_ = current_encoding_;
+    current_encoding_ = type_encoding;
+    return *this;
+}
+
+Cdr& Cdr::xcdr2_end_serialize_type(
+        const Cdr::state& current_state)
+{
+    assert(EncodingAlgorithmFlag::PLAIN_CDR2 == current_encoding_ ||
+            EncodingAlgorithmFlag::DELIMIT_CDR2 == current_encoding_ ||
+            EncodingAlgorithmFlag::PL_CDR2 == current_encoding_);
+    if (EncodingAlgorithmFlag::PLAIN_CDR2 != current_encoding_)
+    {
+        auto last_offset = offset_;
+        set_state(current_state);
+        const size_t member_serialized_size = last_offset - offset_ /*DHEADER ->*/ - 4 - alignment(4);
+        serialize(static_cast<uint32_t>(member_serialized_size));
+        jump(member_serialized_size);
+        serialized_member_size_ = SERIALIZED_MEMBER_SIZE;
+    }
+    current_encoding_ = current_state.previous_encoding_;
+    return *this;
+}
+
+Cdr& Cdr::xcdr1_deserialize_type(
+        EncodingAlgorithmFlag type_encoding,
+        std::function<bool (Cdr&, const MemberId&)> functor)
+{
+    assert(EncodingAlgorithmFlag::PLAIN_CDR == type_encoding ||
+            EncodingAlgorithmFlag::PL_CDR == type_encoding);
+    assert(offset_ == m_cdrBuffer.begin() ? current_encoding_ == type_encoding : true);
+    Cdr::state current_state(*this);
+
+    if (EncodingAlgorithmFlag::PL_CDR == type_encoding)
+    {
+        while (xcdr1_deserialize_member_header(next_member_id_, current_state))
+        {
+            bool deser_value = functor(*this, next_member_id_);
+
+            if (!deser_value)
+            {
+                if (next_member_id_.must_understand)
+                {
+                    throw BadParamException("Cannot deserialize a member with flag must_understand");
+                }
+                else
+                {
+                    jump(current_state.member_size_);
+                }
+            }
+
+            if (current_state.member_size_ != offset_ - origin_)
+            {
+                throw BadParamException("Member size provided by member header is not equal at the real decoded member");
```

Comment:
```suggestion
                throw BadParamException("Member size provided by member header is not equal to the real decoded member size");
```

Please, apply to similar BadParamException message below too.

Reply:
Done in 6ade16d